tests/loggen/loggen.c

```diff
@@ -609,6 +600,49 @@ gint active_finished;
 gint connect_finished;
 guint64 sum_count;
 
+/* Multithreaded OpenSSL */
+GMutex **ssl_lock_cs;
+
+static void
+ssl_threadid_callback(CRYPTO_THREADID *tid)
+{
+  CRYPTO_THREADID_set_pointer(tid, g_thread_self());
+}
+
+static void
+ssl_locking_callback(int mode, int n, const char *file, int line)
+{
+  if (mode & CRYPTO_LOCK)
+    g_mutex_lock(ssl_lock_cs[n]);
+  else
+    g_mutex_unlock(ssl_lock_cs[n]);
+}
+
+static void
+ssl_thread_setup(void)
+{
+  int i;
+  ssl_lock_cs = g_malloc(CRYPTO_num_locks() * sizeof(GMutex *));
+
+  for (i = 0; i < CRYPTO_num_locks(); ++i)
+    ssl_lock_cs[i] = g_mutex_new();
+
+  CRYPTO_THREADID_set_callback(ssl_threadid_callback);
+  CRYPTO_set_locking_callback(ssl_locking_callback);
+}
+
+static void
+ssl_thread_cleanup(void)
+{
+  int i;
+  CRYPTO_set_locking_callback(NULL);
+
+  for (i = 0; i < CRYPTO_num_locks(); ++i)
+    g_mutex_free(ssl_lock_cs[i]);
+
+  g_free(ssl_lock_cs);
+}
+
 gpointer
```

Comment:
Thanks @bazsi, you are right. I just noticed that the same lock functionality exists in `lib/crypto.c`.
Maybe I could link against `lib/crypto` statically and use `crypto_init_threading()` instead. What do you think?